src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,875 @@
+   *
+   * \param [in] ui The index of the weight in on the first axis
+   * \param [in] vi The index of the weight in on the second axis
+   * \param [in] weight The updated value of the weight
+   * \pre Requires that the surface be rational
+   * \pre Requires that the weight be positive
+   */
+  void setWeight(int ui, int vi, T weight)
+  {
+    SLIC_ASSERT(isRational());
+    SLIC_ASSERT(weight > 0);
+
+    m_weights(ui, vi) = weight;
+  };
+
+  /// Checks equality of two Bezier Patches
+  friend inline bool operator==(const BezierPatch<T, NDIMS>& lhs,
+                                const BezierPatch<T, NDIMS>& rhs)
+  {
+    return (lhs.m_controlPoints == rhs.m_controlPoints) &&
+      (lhs.m_weights == rhs.m_weights);
+  }
+
+  friend inline bool operator!=(const BezierPatch<T, NDIMS>& lhs,
+                                const BezierPatch<T, NDIMS>& rhs)
+  {
+    return !(lhs == rhs);
+  }
+
+  /// Returns a copy of the Bezier patch's control points
+  CoordsMat getControlPoints() const { return m_controlPoints; }
+
+  /// Returns a copy of the Bezier patch's weights
+  axom::Array<T, 2> getWeights() const { return m_weights; }
+
+  /*!
+   * \brief Reverses the order of one direction of the Bezier patch's control points and weights
+   *
+   * \param [in] axis orientation of patch. 0 to reverse in u, 1 for reverse in v
+   */
+  void reverseOrientation(int axis)
+  {
+    const int ord_u = getOrder_u();
+    const int mid_u = (ord_u + 1) / 2;
+
+    const int ord_v = getOrder_v();
+    const int mid_v = (ord_v + 1) / 2;
+
+    if(axis == 0)
+    {
+      for(int q = 0; q <= ord_v; ++q)
+      {
+        for(int i = 0; i < mid_u; ++i)
+          axom::utilities::swap(m_controlPoints(i, q),
+                                m_controlPoints(ord_u - i, q));
+
+        if(isRational())
+          for(int i = 0; i < mid_u; ++i)
+            axom::utilities::swap(m_weights(i, q), m_weights(ord_u - i, q));
+      }
+    }
+    else
+    {
+      for(int p = 0; p <= ord_u; ++p)
+      {
+        for(int i = 0; i < mid_v; ++i)
+          axom::utilities::swap(m_controlPoints(p, i),
+                                m_controlPoints(p, ord_v - i));
+
+        if(isRational())
+          for(int i = 0; i < mid_v; ++i)
+            axom::utilities::swap(m_weights(p, i), m_weights(p, ord_v - i));
+      }
+    }
+  }
+
+  /// Swap the axes such that s(u, v) becomes s(v, u)
+  void swapAxes()
+  {
+    const int ord_u = getOrder_u();
+    const int ord_v = getOrder_v();
+
+    CoordsMat new_controlPoints(ord_v + 1, ord_u + 1);
+
+    for(int p = 0; p <= ord_u; ++p)
+      for(int q = 0; q <= ord_v; ++q)
+        new_controlPoints(q, p) = m_controlPoints(p, q);
+
+    m_controlPoints = new_controlPoints;
+
+    if(isRational())
+    {
+      axom::Array<T, 2> new_weights(ord_v + 1, ord_u + 1);
+      for(int p = 0; p <= ord_u; ++p)
+        for(int q = 0; q <= ord_v; ++q) new_weights(q, p) = m_weights(p, q);
+
+      m_weights = new_weights;
```

Comment:
Possible optimization: Can we do this in place instead of using a copy? This might depend on how flexible `axom::Array<T,2>` is about reshaping to an array using the same underlying buffer.

Reply:
I looked around in the docs for `axom::Array` for some kind of reshape method, but I don't think there is one. There's `resize` of course, but it isn't clear to me what that does to the underlying data.

Moreover, the operation we actually would want is a transpose, swapping the data at `m[i,j]` and `m[j,i]` rather than just swapping the spans in each direction. If the array was square, then you could definitely do it in place with a set of `swap` operations, but I'd have to think about how to accomplish the same thing for rectangular arrays.

Reply:
@publixsubfan, @ltaylor16 or others -- does `axom::Array` have a need for a more general reshape that changes the 2D (or higher) dimensions of the array?